client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonDecodingUtils.java

```diff
@@ -0,0 +1,460 @@
+import static io.trino.client.ClientStandardTypes.JSON;
+import static io.trino.client.ClientStandardTypes.MAP;
+import static io.trino.client.ClientStandardTypes.REAL;
+import static io.trino.client.ClientStandardTypes.ROW;
+import static io.trino.client.ClientStandardTypes.SMALLINT;
+import static io.trino.client.ClientStandardTypes.SPHERICAL_GEOGRAPHY;
+import static io.trino.client.ClientStandardTypes.TIME;
+import static io.trino.client.ClientStandardTypes.TIMESTAMP;
+import static io.trino.client.ClientStandardTypes.TIMESTAMP_WITH_TIME_ZONE;
+import static io.trino.client.ClientStandardTypes.TIME_WITH_TIME_ZONE;
+import static io.trino.client.ClientStandardTypes.TINYINT;
+import static io.trino.client.ClientStandardTypes.UUID;
+import static io.trino.client.ClientStandardTypes.VARCHAR;
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public final class JsonDecodingUtils
+{
+    private JsonDecodingUtils() {}
+
+    static TypeDecoder[] createTypeDecoders(List<Column> columns)
+    {
+        requireNonNull(columns, "columns is null");
+        TypeDecoder[] decoders = new TypeDecoder[columns.size()];
+        int index = 0;
+        for (Column column : columns) {
+            decoders[index++] = createTypeDecoder(column.getTypeSignature());
+        }
+        return decoders;
+    }
+
+    interface TypeDecoder
+    {
+        Object decode(JsonParser parser)
+                throws IOException;
+    }
+
+    private static TypeDecoder createTypeDecoder(ClientTypeSignature signature)
+    {
+        switch (signature.getRawType()) {
+            case BIGINT:
+                return new BigIntegerDecoder();
+            case INTEGER:
+                return new IntegerDecoder();
+            case SMALLINT:
+                return new SmallintDecoder();
+            case TINYINT:
+                return new TinyintDecoder();
+            case DOUBLE:
+                return new DoubleDecoder();
+            case REAL:
+                return new RealDecoder();
+            case BOOLEAN:
+                return new BooleanDecoder();
+            case ARRAY:
+                return new ArrayDecoder(signature);
+            case MAP:
+                return new MapDecoder(signature);
+            case ROW:
+                return new RowDecoder(signature);
+            case VARCHAR:
+            case JSON:
+            case TIME:
+            case TIME_WITH_TIME_ZONE:
+            case TIMESTAMP:
+            case TIMESTAMP_WITH_TIME_ZONE:
+            case DATE:
+            case INTERVAL_YEAR_TO_MONTH:
+            case INTERVAL_DAY_TO_SECOND:
+            case IPADDRESS:
+            case UUID:
+            case DECIMAL:
+            case CHAR:
+            case GEOMETRY:
+            case SPHERICAL_GEOGRAPHY:
+                return new StringDecoder();
+            default:
```

Comment:
We should encode the physical type in the response metadata so that the client can decode the data even if it doesn't know about the type.

Reply:
I scheduled a meeting on friday to discuss that

Reply:
That's what I was going to propose that we should add an SPI that would describe the logical type using physical, on the wire types (like basic primitives, row, array and map).

Reply:
This is for backward compatibility with the existing format. I'll add a new v2 soon that will be dedicated for spooling protocol.